src/adc.rs

```diff
@@ -0,0 +1,858 @@
+    PA0<Analog> => 0,
+    PA1<Analog> => 1,
+    PA2<Analog> => 2,
+    PA3<Analog> => 3,
+    PA4<Analog> => 4,
+    PA5<Analog> => 5,
+    PA6<Analog> => 6,
+    PA7<Analog> => 7,
+    PB0<Analog> => 8,
+    PB1<Analog> => 9,
+    PC0<Analog> => 10,
+    PC1<Analog> => 11,
+    PC2<Analog> => 12,
+    PC3<Analog> => 13,
+    PC4<Analog> => 14,
+    PC5<Analog> => 15,
+    Temperature<Enabled> => 16,
+    Vrefint<Enabled> => 17,
+);
+
+adc_pins!(ADC1,
+    PA0<Analog> => 0,
+    PA1<Analog> => 1,
+    PA2<Analog> => 2,
+    PA3<Analog> => 3,
+    PA4<Analog> => 4,
+    PA5<Analog> => 5,
+    PA6<Analog> => 6,
+    PA7<Analog> => 7,
+    PB0<Analog> => 8,
+    PB1<Analog> => 9,
+    PC0<Analog> => 10,
+    PC1<Analog> => 11,
+    PC2<Analog> => 12,
+    PC3<Analog> => 13,
+    PC4<Analog> => 14,
+    PC5<Analog> => 15,
+);
+
+/// Analog to Digital Converter
+pub struct Adc<ADC, ED> {
+    rb: ADC,
+    config: config::AdcConfig,
+    _enabled: PhantomData<ED>,
+}
+
+// applies to all ADCs
+impl<ADC> Adc<ADC, Disabled> {
+    /// Set ADC clock div
+    pub fn set_clock(&mut self, clock: config::Clock, rcu: &mut Rcu) {
+        use self::config::Clock::*;
+
+        match clock {
+            Apb2_div_2 | Apb2_div_4 | Apb2_div_6 | Apb2_div_8 => unsafe {
+                rcu.regs
+                    .cfg0
+                    .modify(|_, w| w.adcpsc_1_0().bits(clock as u8).adcpsc_2().clear_bit());
+            },
+            Apb2_div_12 | Apb2_div_16 => unsafe {
+                rcu.regs.cfg0.modify(|_, w| {
+                    w.adcpsc_1_0()
+                        .bits(clock as u8 >> 2)
+                        .adcpsc_2()
+                        .bit(clock as u8 & 0x1 == 0x1)
+                });
+            },
+        }
+    }
+
+    /// Applies all fields in AdcConfig
+    pub fn apply_config(&mut self, config: config::AdcConfig) {
+        self.config = config;
+    }
+
+    /// Sets the sampling resolution
+    pub fn set_resolution(&mut self, resolution: config::Resolution) {
+        self.config.resolution = resolution;
+    }
+
+    /// Sets the DR register alignment to left or right
+    pub fn set_align(&mut self, align: config::Align) {
+        self.config.align = align;
+    }
+
+    /// Enables and disables scan mode
+    pub fn set_scan(&mut self, scan: config::Scan) {
+        self.config.scan = scan;
+    }
+
+    /// Enables and disables continuous mode
+    pub fn set_continuous(&mut self, continuous: config::Continuous) {
+        self.config.continuous = continuous;
+    }
+}
+
+impl Adc<ADC0, Disabled> {
+    // adc_mode_config for single channel, use free mode
+    fn pre_configure(&mut self, resolution: config::Resolution) {
+        // resolution for ADC0
+        unsafe {
+            self.rb
+                .ovsampctl
+                .modify(|_, w| w.dres().bits(resolution as u8));
+        }
+    }
+}
+
+impl Adc<ADC1, Disabled> {
+    fn pre_configure(&mut self, _resolution: config::Resolution) {}
```

Comment:
Why there is no `ovsampctl` register in ADC1 ?

Reply:
https://docs.rs/gd32vf103-pac/0.4.0/gd32vf103_pac/adc1/struct.RegisterBlock.html